extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,211 @@
+import org.apache.druid.data.input.MapBasedInputRow;
+import org.apache.druid.data.input.kafka.KafkaRecordEntity;
+import org.apache.druid.java.util.common.CloseableIterators;
+import org.apache.druid.java.util.common.Pair;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.common.parsers.CloseableIterator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class KafkaInputReader implements InputEntityReader
+{
+  private static final Logger log = new Logger(KafkaInputReader.class);
+  private static final String DEFAULT_KEY_STRING = "key";
+  private static final String DEFAULT_TIMESTAMP_STRING = "timestamp";
+
+  private final InputRowSchema inputRowSchema;
+  private final KafkaRecordEntity record;
+  private final KafkaHeaderReader headerParser;
+  private final InputEntityReader keyParser;
+  private final InputEntityReader valueParser;
+  private final String keyLabelPrefix;
+  private final String recordTimestampLabelPrefix;
+
+  public KafkaInputReader(
+      InputRowSchema inputRowSchema,
+      KafkaRecordEntity record,
+      KafkaHeaderReader headerParser,
+      InputEntityReader keyParser,
+      InputEntityReader valueParser,
+      String keyLabelPrefix,
+      String recordTimestampLabelPrefix
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser; //Header parser can be null by config
+    this.keyParser = keyParser; //Key parser can be null by config and data
+    this.valueParser = valueParser; //value parser can be null by data (tombstone records)
+    this.keyLabelPrefix = keyLabelPrefix;
+    this.recordTimestampLabelPrefix = recordTimestampLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    // Add kafka record timestamp to the mergelist
+    List<Pair<String, Object>> mergeList = new ArrayList<>();
+    if (headerParser != null) {
+      List<Pair<String, Object>> headerList = headerParser.read();
+      mergeList.addAll(headerList);
+    }
+
+    Pair ts = new Pair(recordTimestampLabelPrefix + DEFAULT_TIMESTAMP_STRING,
+                      record.getRecord().timestamp());
+    if (!mergeList.contains(ts)) {
+      mergeList.add(ts);
+    }
+
+    if (keyParser != null) {
+      try (CloseableIterator<InputRow> keyIterator = keyParser.read()) {
+        // Key currently only takes the first row and ignores the rest.
+        if (keyIterator.hasNext()) {
+          // Return type for the key parser should be of type MapBasedInputRow
+          // Parsers returning other types are not compatible currently.
+          MapBasedInputRow keyRow = (MapBasedInputRow) keyIterator.next();
+          Pair key = new Pair(keyLabelPrefix + DEFAULT_KEY_STRING,
```

Comment:
same here for the column name, we should compute the final column name once per input format instance and then pass it down to the reader to avoid unnecessary concatenation